app/lib/backend/storage_rewrite.rb

```diff
@@ -1,106 +1,135 @@
+        def rewrite(class_name, ids)
+          rewriter = Backend::StorageRewrite.const_get("#{class_name}Rewriter")
 
-      index = 0
-      progress = -> do
-        percent = ((index + 1) / total_count.to_f) * 100.0
-        yield percent if block_given?
-        index += 1
+          rewriter.rewrite(ids)
+        end
       end
 
-      services.find_each do |service|
-        service.update_backend_service
-        progress.call
-
-        service.service_tokens.find_each(&ServiceTokenService.method(:update_backend))
+      # All accounts eligible for backend sync
+      # Optionally include inactive (deleted and suspended)
+      def providers
+        @providers ||=
+          begin
+            providers = Account.providers_with_master
+            providers = providers.without_deleted.without_suspended unless include_inactive
+            providers
+          end
       end
 
-      metrics.find_each do |metric|
-        metric.sync_backend!
-        progress.call
+      # Schedule all objects for all providers, or execute inline
+      def rewrite_all
+        providers.each do |account|
+          rewrite_provider(account.id)
+        end
+        nil
       end
 
-      usage_limits.find_each do |usage_limit|
-        usage_limit.send(:update_backend_usage_limit)
-        progress.call
+      # Schedule a single provider or execute inline
+      def rewrite_provider(id)
+        logger.info "Rewriting backend storage for provider #{id}..."
+
+        provider = providers.find_by(id: id)
+        unless provider
+          logger.error("Provider with ID #{id} not found")
+          return
+        end
+
+        logger.info "Rewriting services for provider #{id}..."
+        process_in_batches(provider.services)
+
+        logger.info "Rewriting buyer applications for provider #{id}..."
+        process_in_batches(provider.buyer_applications)
+
+        logger.info "Rewriting provider applications for provider #{id}..."
+        process_in_batches(provider.bought_cinstances)
+
+        logger.info "Rewriting metrics for provider #{id}..."
+        process_in_batches(Metric.by_provider(provider))
+
+        logger.info "Rewriting usage limits for provider #{id}..."
+        process_in_batches(provider.usage_limits)
       end
 
-      update_cinstance = ->(cinstance) do
-        cinstance.send(:update_backend_application)
-        cinstance.send(:update_backend_user_key_to_application_id_mapping)
-        if cinstance.provider_account
-          cinstance.application_keys.each { |k| k.send(:update_backend_value) }
-          cinstance.referrer_filters.each { |f| f.send(:update_backend_value) }
+      # Process each collection in batches of a predefined size
+      # Either schedule async execution (if :enqueuer opt is provided), or execute inline
+      def process_in_batches(collection)
+        collection.in_batches(of: BATCH_SIZE) do |batch|
+          ids = batch.pluck(:id)
+
+          next if ids.empty?
```

Comment:
Can this happen? a batch be empty but still be a next batch?

Reply:
Removed this.